Add charm-architecture.md (#69)
canonical/pollen-operator: docs/explanation/charm-architecture.md

# Charm architecture

> ​The Pollen operator is a machine charm designed to deploy and manage Pollen directly on machines. Pollen is a high-performance, scalable, free web server that provides 'Entropy-as-a-Service' by offering small strings of entropy over both HTTPS and clear-text HTTP connections.

Reviewer: s/directly on machines/on top of VMs and bare metal?

> ​The Pollen operator is a machine charm designed to deploy and manage Pollen directly on machines. Pollen is a high-performance, scalable, free web server that provides 'Entropy-as-a-Service' by offering small strings of entropy over both HTTPS and clear-text HTTP connections.

Reviewer: Can you link to the pollen docs on the first reference to it?

> ​The Pollen operator is a machine charm designed to deploy and manage Pollen directly on machines. Pollen is a high-performance, scalable, free web server that provides 'Entropy-as-a-Service' by offering small strings of entropy over both HTTPS and clear-text HTTP connections.

Reviewer 2: ```suggestion
​The Pollen operator is a machine charm designed to deploy and manage Pollen directly on machines. Pollen is a high-performance, scalable, free web server that provides "Entropy-as-a-Service" by offering small strings of entropy over both HTTPS and clear-text HTTP connections.
```
Nitpick

> ​The [Pollen](https://github.com/dustinkirkland/pollen) operator is a machine charm designed to deploy and manage Pollen directly on VMs and bare metal. Pollen is a high-performance, scalable, free web server that provides "Entropy-as-a-Service" by offering small strings of entropy over both HTTPS and clear-text HTTP connections.

Reviewer 2: ```suggestion
​The Pollen operator is a machine charm designed to deploy and manage Pollen directly on VMs and bare metal. [Pollen](https://github.com/dustinkirkland/pollen) is a high-performance, scalable, free web server that provides "Entropy-as-a-Service" by offering small strings of entropy over both HTTPS and clear-text HTTP connections.
```
Nitpick -- I think it makes more sense to add the Pollen link when describing the software, instead of when you're describing the charm.

> This charm does not use an OCI image, as it runs directly on a machine rather than in a container. The charm uses the [pollen](https://github.com/canonical/pollen/blob/main/snap/snapcraft.yaml) [snap](https://ubuntu.com/core/docs/snaps-in-ubuntu-core) to install pollen during deployment.

Reviewer 2: ```suggestion
This charm does not use an OCI image, as it runs directly on a machine rather than in a container. The charm uses the [pollen](https://github.com/canonical/pollen/blob/main/snap/snapcraft.yaml) [snap](https://ubuntu.com/core/docs/snaps-in-ubuntu-core) to install Pollen during deployment.
```

> This charm does not use an OCI image, as it runs directly on a machine rather than in a container. Pollen is implemented in Golang and provides small random strings to its clients over network connections. It utilizes TLS (SSL) to ensure privacy, security, and non-repudiation of connections among its clients. Canonical provides a Pollen server as a service to the Ubuntu community at [https://entropy.ubuntu.com](https://entropy.ubuntu.com).

Reviewer: I think "This charm does not use an OCI image, as it runs directly on a machine rather than in a container. Pollen is implemented in Golang and provides small random strings to its clients over network connections. It utilizes TLS (SSL) to ensure privacy, security, and non-repudiation of connections among its clients." is a bit redundant bit what's mentoned above. I'd remove this sentence

Reviewer 2: I would keep "This charm does not use an OCI image, as it runs directly on a machine rather than in a container." even if it is redundant. I think it's good to clarify that there's no OCI image (it may be clear to seasoned charmers, but might not be as clear to new developers).

Author: removed

## Metrics

> Pollen exposes the following metrics, which are visible in Grafana after [integrating with COS](https://charmhub.io/pollen/docs/how-to-relate-to-cos).

Author: done

> 1. **pollen_http_requests_total** - Total HTTP requests
> 2. **pollen_http_responses_codes** - Total HTTP requests by code
> 3. **pollen_system_entropy** - System entropy
> 4. **pollen_http_response_seconds** - HTTP response time by code
> 5. **pollen_response_entropy_per_byte** - Entropy per byte
> 6. **pollen_response_entropy_arithmetic_mean_deviation** - Entropy arithmetic mean deviation

Reviewer 2: ```suggestion
1. **pollen_http_requests_total**: Total HTTP requests
2. **pollen_http_responses_codes**: Total HTTP requests by code
3. **pollen_system_entropy**: System entropy
4. **pollen_http_response_seconds**: HTTP response time by code
5. **pollen_response_entropy_per_byte**: Entropy per byte
6. **pollen_response_entropy_arithmetic_mean_deviation**: Entropy arithmetic mean deviation
```
Nitpick

## Juju events

> According to the [Juju SDK](https://canonical-juju.readthedocs-hosted.com/en/latest/user/reference/hook/): "an event is a data structure that encapsulates part of the execution context of a charm".

Reviewer: Do we really need this statement?

Reviewer 2: I agree, this sentence can be removed. If you would like to add a link, I suggest:

```
> See more in the Juju docs: [Hook](https://documentation.ubuntu.com/juju/latest/user/reference/hook/)
```

Author: This statement was part of the charm-architecture template. I think it would be useful for first-time users who don't understand the concept of hooks.

Author: Changing this line to what Erin mentioned (https://github.com/canonical/pollen-operator/pull/69#discussion_r2031305990)

2. [upgrade-charm](https://documentation.ubuntu.com/juju/latest/reference/hook/index.html#upgrade-charm): Runs once immediately after the charm directory contents have been changed by an unforced charm upgrade operation, and may do so after a forced upgrade; but will not be run after a forced upgrade from an existing error state. **Action**: Install packages and write configuration files for pollen.
3. [start](https://documentation.ubuntu.com/juju/latest/reference/hook/index.html#start): Triggered when a unit's initialization is complete. **Action**: Start the pollen service.
4. [stop](https://documentation.ubuntu.com/juju/latest/reference/hook/index.html#stop): Triggered when a Juju controller is ready to destroy the unit. **Action**: Stop the pollen service.
5. [website-relation-changed](https://documentation.ubuntu.com/juju/latest/reference/hook/index.html#endpoint-relation-changed): Emitted when another unit involved in the website relation (from either side) touches the relation data. **Action**: Updates the relation data with the website details.

> See more in the Juju docs: [Hook](https://documentation.ubuntu.com/juju/latest/user/reference/hook/)

Reviewer 2: ```suggestion
> See more in the Juju docs: [Hook](https://documentation.ubuntu.com/juju/latest/user/reference/hook/)
```
Nitpick
